### PR TITLE
Module Core Refactor

### PR DESCRIPTION
**Pull request by GWSzeto**

### CODE REVIEW

File: src/core/token/ERC1155Core.sol [hunk 1]
```diff
@@ -167,15 +196,57 @@ contract ERC1155Core is ERC1155, Core, Multicallable {
...
+    function mintWithSignature(
+        address to,
+        uint256 tokenId,
+        uint256 value,
+        string calldata baseURI,
+        bytes calldata data,
+        bytes memory signature
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(abi.encode(TYPEHASH_SIGNATURE_MINT_ERC1155, to, tokenId, value, keccak256(bytes(baseURI)), data))
```
Comment (GWSzeto) [inline src/core/token/ERC1155Core.sol:234]: keccack256 the data part here

File: src/core/token/ERC1155Core.sol [hunk 2]
```diff
@@ -167,15 +196,57 @@ contract ERC1155Core is ERC1155, Core, Multicallable {
...
+        ).recover(signature);
+
+        if (!OwnableRoles(address(this)).hasAllRoles(signer, Role._MINTER_ROLE)) {
```
Comment (GWSzeto) [inline src/core/token/ERC1155Core.sol:237]: remove the `OwnableRoles(address(this))` part as it's already the core contract

File: src/core/token/ERC1155CoreInitializable.sol [hunk 3]
```diff
@@ -172,15 +202,57 @@ contract ERC1155CoreInitializable is ERC1155, Core, Multicallable, Initializable
...
+    function mintWithSignature(
+        address to,
+        uint256 tokenId,
+        uint256 value,
+        string calldata baseURI,
+        bytes calldata data,
+        bytes memory signature
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(abi.encode(TYPEHASH_SIGNATURE_MINT_ERC1155, to, tokenId, value, keccak256(bytes(baseURI)), data))
```
Comment (GWSzeto) [inline src/core/token/ERC1155CoreInitializable.sol:240]: keccack256(data)

File: src/core/token/ERC721Core.sol [hunk 4]
```diff
@@ -149,8 +179,44 @@ contract ERC721Core is ERC721AQueryable, Core, Multicallable {
...
+        uint256 quantity,
+        string calldata baseURI,
+        bytes calldata data,
+        bytes memory signature
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(abi.encode(TYPEHASH_SIGNATURE_MINT_ERC721, to, quantity, keccak256(bytes(baseURI)), data))
+        ).recover(signature);
+
+        if (!OwnableRoles(address(this)).hasAllRoles(signer, Role._MINTER_ROLE)) {
```
Comment (GWSzeto) [inline src/core/token/ERC721Core.sol:210]: remove OwnableRoles(address(this))

File: src/core/token/ERC721Core.sol [hunk 5]
```diff
@@ -149,8 +179,44 @@ contract ERC721Core is ERC721AQueryable, Core, Multicallable {
...
+     */
+    function mintWithSignature(
+        address to,
+        uint256 quantity,
+        string calldata baseURI,
+        bytes calldata data,
+        bytes memory signature
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(abi.encode(TYPEHASH_SIGNATURE_MINT_ERC721, to, quantity, keccak256(bytes(baseURI)), data))
```
Comment (GWSzeto) [inline src/core/token/ERC721Core.sol:207]: keccack256(Data)

File: src/core/token/ERC721CoreInitializable.sol [hunk 6]
```diff
@@ -169,8 +191,44 @@ contract ERC721CoreInitializable is ERC721AQueryableUpgradeable, Core, Multicall
...
+     */
+    function mintWithSignature(
+        address to,
+        uint256 quantity,
+        string calldata baseURI,
+        bytes calldata data,
+        bytes memory signature
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(abi.encode(TYPEHASH_SIGNATURE_MINT_ERC721, to, quantity, keccak256(bytes(baseURI)), data))
```
Comment (GWSzeto) [inline src/core/token/ERC721CoreInitializable.sol:219]: keccack256(data)

File: src/core/token/ERC721CoreInitializable.sol [hunk 7]
```diff
@@ -169,8 +191,44 @@ contract ERC721CoreInitializable is ERC721AQueryableUpgradeable, Core, Multicall
...
+        ).recover(signature);
+
+        if (!OwnableRoles(address(this)).hasAllRoles(signer, Role._MINTER_ROLE)) {
```
Comment (GWSzeto) [inline src/core/token/ERC721CoreInitializable.sol:222]: remove

File: src/core/token/ERC1155CoreInitializable.sol [hunk 8]
```diff
@@ -172,15 +202,57 @@ contract ERC1155CoreInitializable is ERC1155, Core, Multicallable, Initializable
...
+        ).recover(signature);
+
+        if (!OwnableRoles(address(this)).hasAllRoles(signer, Role._MINTER_ROLE)) {
```
Comment (GWSzeto) [inline src/core/token/ERC1155CoreInitializable.sol:243]: move to module

File: src/module/token/metadata/BatchMetadataERC721.sol [hunk 9]
```diff
@@ -131,20 +151,8 @@ contract BatchMetadataERC721 is Module {
...
-
-        uint256 rangeStart = _batchMetadataStorage().nextTokenIdRangeStart;
-        uint256 rangeEndNonInclusive = rangeStart + _amount;
-
-        _batchMetadataStorage().nextTokenIdRangeStart = rangeEndNonInclusive;
-        _batchMetadataStorage().tokenIdRangeEnd.push(rangeEndNonInclusive);
-        _batchMetadataStorage().baseURIOfTokenIdRange[rangeEndNonInclusive] = _baseURI;
-
-        emit NewMetadataBatch(rangeStart, rangeEndNonInclusive, _baseURI);
-        emit BatchMetadataUpdate(rangeStart, rangeEndNonInclusive - 1);
```
Comment (GWSzeto) [inline src/module/token/metadata/BatchMetadataERC721.sol:147]: Look into the spec around this event `UpdateMetadata` spec

File: src/module/token/minting/ClaimableERC1155.sol [hunk 10]
```diff
@@ -37,7 +38,13 @@ library ClaimableStorage {
 
 }
 
-contract ClaimableERC1155 is Module, EIP712, BeforeMintCallbackERC1155, IInstallationCallback {
+contract ClaimableERC1155 is
+    Module,
+    EIP712,
```
Comment (GWSzeto) [inline src/module/token/minting/ClaimableERC1155.sol:43]: Remove this

File: src/core/token/ERC1155Core.sol [hunk 11]
```diff
@@ -167,15 +196,57 @@ contract ERC1155Core is ERC1155, Core, Multicallable {
      *  @param to The address to mint the token to.
      *  @param tokenId The tokenId to mint.
      *  @param value The amount of tokens to mint.
+     *  @param baseURI The base URI for the token metadata.
      *  @param data ABI encoded data to pass to the beforeMint hook.
      */
-    function mint(address to, uint256 tokenId, uint256 value, bytes memory data) external payable {
+    function mint(address to, uint256 tokenId, uint256 value, string calldata baseURI, bytes memory data)
```
Comment (GWSzeto) [inline src/core/token/ERC1155Core.sol:202]: rename to either `amount` or `quantity`, just make the naming consistent

File: src/module/token/minting/ClaimableERC1155.sol [hunk 12]
```diff
@@ -86,11 +93,8 @@ contract ClaimableERC1155 is Module, EIP712, BeforeMintCallbackERC1155, IInstall
      *  @param uid A unique identifier for the minting request.
      */
     struct ClaimRequestERC1155 {
```
Comment (GWSzeto) [inline src/module/token/minting/ClaimableERC1155.sol:95]: ClaimSignatureParams

File: src/core/token/ERC1155Core.sol [hunk 13]
```diff
@@ -1,7 +1,11 @@
 // SPDX-License-Identifier: Apache-2.0
 pragma solidity ^0.8.20;
 
+import "forge-std/console.sol";
```
Comment (joaquim-verges) [inline src/core/token/ERC1155Core.sol:4]: remove

File: src/core/token/ERC1155Core.sol [hunk 14]
```diff
@@ -166,29 +195,75 @@ contract ERC1155Core is ERC1155, Core, Multicallable {
...
+        }
+        _beforeMint(to, tokenId, amount, data);
 
-        _totalSupply[tokenId] += value;
-        _mint(to, tokenId, value, "");
+        _totalSupply[tokenId] += amount;
+        _mint(to, tokenId, amount, "");
+    }
+
+    receive() external payable {}
```
Comment (joaquim-verges) [inline src/core/token/ERC1155Core.sol:215]: remove

File: src/core/token/ERC1155Core.sol [hunk 15]
```diff
@@ -166,29 +195,75 @@ contract ERC1155Core is ERC1155, Core, Multicallable {
...
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(
+                abi.encode(
+                    TYPEHASH_SIGNATURE_MINT_ERC1155, to, tokenId, amount, keccak256(bytes(baseURI)), keccak256(data)
+                )
+            )
+        ).recover(signature);
+        console.log("signer");
+        console.logAddress(signer);
```
Comment (joaquim-verges) [inline src/core/token/ERC1155Core.sol:243]: cleanup

File: src/core/token/ERC1155Core.sol [hunk 16]
```diff
@@ -166,29 +195,75 @@ contract ERC1155Core is ERC1155, Core, Multicallable {
...
+        bytes calldata data,
+        bytes memory signature
+    ) external payable {
+        address signer = _hashTypedData(
+            keccak256(
+                abi.encode(
+                    TYPEHASH_SIGNATURE_MINT_ERC1155, to, tokenId, amount, keccak256(bytes(baseURI)), keccak256(data)
+                )
+            )
+        ).recover(signature);
```
Comment (joaquim-verges) [inline src/core/token/ERC1155Core.sol:241]: you're not checking that the signer has minter role here?

File: src/callback/BeforeMintWithSignatureCallbackERC1155.sol [hunk 17]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+contract BeforeMintWithSignatureCallbackERC1155 {
```
Comment (joaquim-verges) [inline src/callback/BeforeMintWithSignatureCallbackERC1155.sol:4]: kinda weird that these are not abstract/interfaces

**Review by joaquim-verges — APPROVED** [review]

**Pull request merged**